src/enzyme_ad/jax/Passes/MarkFunctionMemoryEffectsPass.cpp

```diff
@@ -0,0 +1,216 @@
+#include "src/enzyme_ad/jax/Passes/Passes.h"
+
+#include "mlir/Analysis/CallGraph.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace enzyme {
+#define GEN_PASS_DEF_MARKFUNCTIONMEMORYEFFECTSPASS
+#include "src/enzyme_ad/jax/Passes/Passes.h.inc"
+} // namespace enzyme
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::enzyme;
+
+namespace {
+
+enum class VisitState { NotVisited, Visiting, Visited };
+
+struct MarkFunctionMemoryEffectsPass
+    : public enzyme::impl::MarkFunctionMemoryEffectsPassBase<
+          MarkFunctionMemoryEffectsPass> {
+  using Base::Base;
+
+  DenseMap<CallGraphNode *, VisitState> visitState;
+  SmallVector<CallGraphNode *> topoOrder;
+
+  bool dfs(CallGraphNode *node) {
+    auto it = visitState.find(node);
+    if (it != visitState.end()) {
+      if (it->second == VisitState::Visiting)
+        return true; // cycle detected
+      return false;
+    }
+    visitState[node] = VisitState::Visiting;
+    for (auto &edge : *node) {
+      CallGraphNode *target = edge.getTarget();
+      if (dfs(target))
+        return true;
+    }
+    visitState[node] = VisitState::Visited;
+    topoOrder.push_back(node);
+    return false;
+  }
+
+  void insertMemoryEffectsAsStringRefs(
+      llvm::SmallDenseSet<StringRef> &effects,
```

Comment:
We can use `MemoryEffects::Effect` instead of StringRef here and only convert when creating the attribute. This avoids strings flying around in the logic. A micro-optimization on top: use `uint8_t` with bit flags for each effect, but that may be an overkill.